doc/sphinx/proofs/automatic-tactics/auto.rst

```diff
@@ -441,80 +425,140 @@ automatically created.
          .. coqtop:: all
 
             Require Import List.
-            Hint Extern 5 ({?X1 = ?X2} + {?X1 <> ?X2}) => generalize  X1, X2; decide equality : eqdec.
+            Hint Extern 5 ({?X1 = ?X2} + {?X1 <> ?X2}) =>
+              generalize  X1, X2; decide equality : eqdec.
             Goal forall a b:list (nat * nat), {a = b} + {a <> b}.
-            Info 1 auto with eqdec.
-
-   .. cmdv:: Hint Cut @regexp : @ident
-      :name: Hint Cut
-
-      .. warning::
+            info_auto.
 
-         These hints currently only apply to typeclass proof search and the
-         :tacn:`typeclasses eauto` tactic.
+   .. cmd:: Hint Cut [ @hints_regexp ] {? : {+ @ident } }
 
-      This command can be used to cut the proof-search tree according to a regular
-      expression matching paths to be cut. The grammar for regular expressions is
-      the following. Beware, there is no operator precedence during parsing, one can
-      check with :cmd:`Print HintDb` to verify the current cut expression:
+      .. DISABLED insertprodn hints_regexp hints_regexp
 
       .. prodn::
-         regexp ::= @ident   (hint or instance identifier)
+         hints_regexp ::= {+ @qualid }   (hint or instance identifier)
          | _   (any hint)
-         | @regexp | @regexp   (disjunction)
-         | @regexp @regexp   (sequence)
-         | @regexp *   (Kleene star)
+         | @hints_regexp | @hints_regexp   (disjunction)
+         | @hints_regexp @hints_regexp   (sequence)
+         | @hints_regexp *   (Kleene star)
          | emp   (empty)
          | eps   (epsilon)
-         | ( @regexp )
+         | ( @hints_regexp )
+
+      Used to cut the proof search tree according to a regular
+      expression that matches the paths to be cut.
+
+      For each hint :n:`@qualid` in the hint database, the current path `p`
+      extended with :n:`@qualid`
+      is matched against the current cut expression `c` associated with the
+      hint database.  If the match succeeds the hint is *not* applied.
+
+      :n:`Hint Cut @hints_regexp` sets the cut expression
+      to :n:`c | regexp`.  The initial cut expression is `emp`.
 
-      The `emp` regexp does not match any search path while `eps`
-      matches the empty path. During proof search, the path of
-      successive successful hints on a search branch is recorded, as a
-      list of identifiers for the hints (note that :cmd:`Hint Extern`\’s do not have
+      The output of :cmd:`Print HintDb` shows the cut expression.
+
+      During proof search, the path of
+      successive successful hints on a search branch is recorded as a
+      list of identifiers for the hints (note that :cmd:`Hint Extern`\s do not have
       an associated identifier).
```

Comment:
This paragraph is useful but it would be better if it were just before the second paragraph ("For each hint qualid in the hint database, the current path p extended with qualid is matched against the current cut expression c associated with the hint database. If the match succeeds the hint is not applied."). The two could also be merged into a single paragraph.